hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/AbstractMapTable.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.schema.map;
+
+import com.hazelcast.sql.impl.QueryException;
+import com.hazelcast.sql.impl.extract.QueryTargetDescriptor;
+import com.hazelcast.sql.impl.schema.Table;
+import com.hazelcast.sql.impl.schema.TableField;
+import com.hazelcast.sql.impl.schema.TableStatistics;
+
+import java.util.List;
+
+/**
+ * Table backed by IMap or ReplicatedMap.
+ */
+public abstract class AbstractMapTable extends Table {
+
+    private final QueryTargetDescriptor keyDescriptor;
+    private final QueryTargetDescriptor valueDescriptor;
+    private final QueryException exception;
+
+    protected AbstractMapTable(
+        String schemaName,
+        String name,
+        List<TableField> fields,
+        TableStatistics statistics,
+        QueryTargetDescriptor keyDescriptor,
+        QueryTargetDescriptor valueDescriptor
+    ) {
+        super(schemaName, name, fields, statistics);
+
+        this.keyDescriptor = keyDescriptor;
+        this.valueDescriptor = valueDescriptor;
+
+        exception = null;
+    }
+
+    protected AbstractMapTable(String schemaName, String name, QueryException exception) {
+        super(schemaName, name, null, null);
+
+        this.keyDescriptor = null;
+        this.valueDescriptor = null;
+
+        this.exception = exception;
+    }
+
+    @Override
+    public int getFieldCount() {
+        checkException();
+
+        return super.getFieldCount();
+    }
+
+    @Override
+    public <T extends TableField> T getField(int index) {
+        checkException();
+
+        return super.getField(index);
+    }
+
+    @Override
+    public TableStatistics getStatistics() {
+        checkException();
+
+        return super.getStatistics();
+    }
+
+    public QueryTargetDescriptor getKeyDescriptor() {
+        return keyDescriptor;
+    }
+
+    public QueryTargetDescriptor getValueDescriptor() {
+        return valueDescriptor;
```

Comment:
checkException()?

Reply:
`checkException()` is needed to let optimizer know that the table exists, but is not functional. Descriptors are used after the optimization process, so an exception check is not needed here. To summarize, exception check is only needed in the overridden methods.

Reply:
assert exception == null?

Reply:
Maybe, but again - exceptions are needed by Calcite integration, that works with `Table` interface, and never touch extension methods.